File: c++/70.climbing-stairs.cpp
#include <bits/stdc++.h>
using namespace std;

/*
 * @lc app=leetcode id=70 lang=cpp
 *
 * [70] Climbing Stairs
 */

// @lc code=start
class Solution {
   public:
    map<int, int> dp;
    int recurse(int n) {
        if (n == 0) return 0;
        if (n == 1) return 1;
        if (n == 2) return 2;
        if (dp[n]) return dp[n];
        return dp[n] = recurse(n - 1) + recurse(n - 2);
    }
    int climbStairs(int n) {
        int ans = recurse(n);
        return ans;
    }
};
// @lc code=end

File: java/70.climbing-stairs.java
import java.util.Arrays;

/*
 * @lc app=leetcode id=70 lang=java
 *
 * [70] Climbing Stairs
 */

// @lc code=start
class Solution {
    int dp[];

    private int recurse(int i, int n) {
        if (i == n)
            return 1;
        if (dp[i] != -1)
            return dp[i];
        int ways = 0;
        if (i + 1 <= n)
            ways += recurse(i + 1, n);
        if (i + 2 <= n)
            ways += recurse(i + 2, n);
        return dp[i] = ways;
    }

    public int climbStairs(int n) {
        dp = new int[50];
        Arrays.fill(dp, -1);
        return recurse(0, n);
    }
}
// @lc code=end
